### PR TITLE
[RISCV] Support (truncate (smin (smax X, C1), C2)) for vnclipu in combineTruncToVnclip.

### PR DESCRIPTION
**Pull request by topperc**
If the smax removed all negative numbers, then we can treat the smin like a umin.

If the smin and smax are in the other order we can swap them and use a vnclipu as long as the smax constant is smaller than the smin constant.

This is based on similar code from X86's detectUSatPattern.

### CODE REVIEW

**Review by lukel97 — APPROVED** [review]

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -16230,17 +16230,35 @@ static SDValue combineTruncToVnclip(SDNode *N, SelectionDAG &DAG,
...
+
+    // Simple case, V is a UMIN.
+    if (SDValue UMin = MatchMinMax(V, ISD::UMIN, RISCVISD::UMIN_VL, HiC))
+      if (HiC.isMask(VT.getScalarSizeInBits()))
+        return UMin;
+
+    // If we have an SMAX that removes negative numbers first, then we can match
+    // SMIN instead of UMIN.
+    if (SDValue SMin = MatchMinMax(V, ISD::SMIN, RISCVISD::SMIN_VL, HiC))
+      if (SDValue SMax = MatchMinMax(SMin, ISD::SMAX, RISCVISD::SMAX_VL, LoC))
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16246]: If we have a known positive input the SMIN, why is that not being canonicalized to a umin?

**Review by preames — APPROVED** [review]
LGTM w/minor comment.  I agree with Luke's point on naming.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: That's a good question. I also ran these tests through InstCombine and it didn't convert the smin to umin either.

Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: @preames  See https://github.com/llvm/llvm-project/pull/82478 and https://github.com/llvm/llvm-project/pull/88505.

@topperc We did this canonicalization in CVP.

**Pull request merged**